ml/nml.cc

```diff
@@ -0,0 +1,1135 @@
+            host->ts.remaining_ut += remaining_ut;
+
+            switch (training_res) {
+                case TRAINING_RESULT_OK:
+                    host->ts.training_result_ok += 1;
+                    break;
+                case TRAINING_RESULT_INVALID_QUERY_TIME_RANGE:
+                    host->ts.training_result_invalid_query_time_range += 1;
+                    break;
+                case TRAINING_RESULT_NOT_ENOUGH_COLLECTED_VALUES:
+                    host->ts.training_result_not_enough_collected_values += 1;
+                    break;
+                case TRAINING_RESULT_NULL_ACQUIRED_DIMENSION:
+                    host->ts.training_result_null_acquired_dimension += 1;
+                    break;
+                case TRAINING_RESULT_CHART_UNDER_REPLICATION:
+                    host->ts.training_result_chart_under_replication += 1;
+                    break;
+            }
+
+            netdata_mutex_unlock(&host->mutex);
+        }
+
+        worker_is_idle();
+        std::this_thread::sleep_for(std::chrono::microseconds{remaining_ut});
+        worker_is_busy(0);
+    }
+}
+
+static void *train_main(void *arg) {
+    size_t max_elements_needed_for_training = Cfg.max_train_samples * (Cfg.lag_n + 1);
+    tls_data.training_cns = new calculated_number_t[max_elements_needed_for_training]();
+    tls_data.scratch_training_cns = new calculated_number_t[max_elements_needed_for_training]();
+
+    nml_host_t *host = reinterpret_cast<nml_host_t *>(arg);
+    nml_host_train(host);
+    return NULL;
+}
+
+void nml_host_start_training_thread(nml_host_t *host) {
+    if (host->threads_running) {
+        error("Anomaly detections threads for host %s are already-up and running.", rrdhost_hostname(host->rh));
+        return;
+    }
+
+    host->threads_running = true;
+    host->threads_cancelled = false;
+    host->threads_joined = false;
+
+    char tag[NETDATA_THREAD_TAG_MAX + 1];
+
+    snprintfz(tag, NETDATA_THREAD_TAG_MAX, "MLTR[%s]", rrdhost_hostname(host->rh));
+    netdata_thread_create(&host->training_thread, tag, NETDATA_THREAD_OPTION_JOINABLE, train_main, static_cast<void *>(host));
+}
+
+void nml_host_stop_training_thread(nml_host_t *host, bool join) {
+    if (!host->threads_running) {
+        error("Anomaly detections threads for host %s have already been stopped.", rrdhost_hostname(host->rh));
+        return;
+    }
+
+    if (!host->threads_cancelled) {
+        host->threads_cancelled = true;
+
+        // Signal the training queue to stop popping-items
+        nml_queue_signal(host->training_queue);
+        netdata_thread_cancel(host->training_thread);
+    }
+
+    if (join && !host->threads_joined) {
+        host->threads_joined = true;
+        host->threads_running = false;
+
+        delete[] tls_data.training_cns;
+        delete[] tls_data.scratch_training_cns;
+
+        netdata_thread_join(host->training_thread, NULL);
+    }
+}
+
+void *nml_detect_main(void *arg) {
+    UNUSED(arg);
+
```

Comment:
@thiagoftsm add a `sleep(30);` here.

Reply:
Unfortunately this did not fix the issue.
When I removed my `/var/cache/netdata` part of data contradiction we were seeing were removed, this means that I did not have contexts anymore when I required https://localhost:19999/api/v1/contexts?options=all and `m[x]` were removed from my dashboard.